[Decision Log] Add decision log for Autocomplete approach (#2241)
Shopify/polaris-react: documentation/decisions/01 - Using a Compound Component Pattern for Autocomplete.md

# Using a compound component patter to build Autocomplete

## Date

October 3, 2019

## Contributors

- Andrew Musgrave
- Chloe Rice
- Dan Rosenthal
- Daniel Leroux
- Dominic McPhee
- Sara Hill

## Summary

> There is a need for Polaris to provide smaller more reusable components so that Polaris can be used by everyone without needing to fork component. The re-development of the Autocomplete component provided an opportunity for exploration.

Reviewer: ```suggestion
There is a need for Polaris to provide smaller, more composable components so that Polaris can be extended without needing to fork components. The re-development of the Autocomplete component provided an opportunity for exploration of the compound component pattern.
```

Reviewer: This might need one more sentence around why composable components allow greater flexibility. IE: not needing to add a prop for any new configuration, as it can be handled as children. Adding an `Icon` to an `Option` feels like a good example to explain this, since this approach would not require adding/maintaining another prop.

## Solution

> Following an (https://github.com/Shopify/polaris-react/issues/1794)[exploration] we decided to build the Autocomplete using a compound component pattern. We explored various ways to do this:

Reviewer: ```suggestion
Following an [exploration](https://github.com/Shopify/polaris-react/issues/1794), we decided to build the Autocomplete using a compound component pattern. We explored various ways to do this.
```

> 1. Having the parent render its children via methods shared via Context: We opted against this because it defied the purpose of having smaller components and cause the component to renderer every time an option as added the Autocompletes results. For example, 100 results cause for 100 calls to the parent.

Reviewer: ```suggestion
1. Having the child delegate rendering to its parent via methods shared through context

We opted against this for a couple of reasons:
- it defeated the purpose of having smaller components
- it causes the component to re-render every time an option is added (one call to the parent render method for each option)
```

Reviewer: > - it defeated the purpose of having smaller components

This needs more detail. Why did it defeat the purpose of having smaller components?

> 2. Using `cloneElement`: We opted against this approach because since cloning happens at run time and TypeScript happens at compile time the benefits of using TypeScript were lost. The experience also felt less performant most likely due to the cloning at every render.

Reviewer: ```suggestion
2. Using `cloneElement`

We opted against this approach because:
- cloning occurs at run time, while TypeScript runs at compile time, meaning the benefits of using TypeScript were lost
- the experience also felt less performant most likely due to the cloning at every render
```

> 3. Having the list of results as a child of the component and the activator text field as a prop: We chose this approach. Overall the approach felt right, the experience not lagging and renders we're kept to the elements being affected only.

Reviewer: ```suggestion
3. Having the list of results as a child of the component and the activator text field as a prop

We chose this approach. Overall the approach felt right, the experience felt snappy and renders were kept to the elements being affected only.
```

Reviewer: This leaves out how context is used to share state with children. That's an important detail for this approach.

> # Using a compound component patter to build Autocomplete

Reviewer 2: ```suggestion
# Using a compound component pattern to build Autocomplete
```

## Date

October 3, 2019

## Contributors

- Andrew Musgrave
- Chloe Rice
- Dan Rosenthal
- Daniel Leroux
- Dominic McPhee
- Sara Hill

## Summary

There is a need for Polaris to provide smaller, more composable components so that Polaris can be extended without needing to fork components. The re-development of the Autocomplete component provided an opportunity for exploration of the compound component pattern.

## Solution

Following an [exploration](https://github.com/Shopify/polaris-react/issues/1794), we decided to build the Autocomplete using a compound component pattern. We explored various ways to do this.

1. Having the child delegate rendering to its parent via methods shared through context

We opted against this for a couple of reasons:
- it defeated the purpose of having smaller components

> - it causes the component to re-render every time an option is added (one call to the parent render method for each option)

Reviewer 3: I wonder if we can go into more detail with this approach? It'll vary depending on architecture and complexity. A child `re-render` won't necessarily cause siblings to re-render. Here's a [sandbox example](https://codesandbox.io/s/amazing-river-6lomm)

2. Using `cloneElement`

We opted against this approach because:
- cloning occurs at run time, while TypeScript runs at compile time, meaning the benefits of using TypeScript were lost
- the experience also felt less performant most likely due to the cloning at every render

> 3. Having the list of results as a child of the component and the activator text field as a prop

Author: As we discussed we'll need to wire up the context but yes, I'll mention it.

> We chose this approach. Overall the approach felt right, the experience felt snappy and renders were kept to the elements being affected only.

Reviewer 3: It's also very similar to an HTML approach without a lot of magic happening behind the scenes. `Stack` has bit us a couple of times 😬
